chain/chain.go

```diff
@@ -0,0 +1,270 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package chain
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/api/metrics"
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/trace"
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/avalanchego/utils/set"
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/hypersdk/internal/emap"
+	"github.com/ava-labs/hypersdk/internal/workers"
+	"github.com/ava-labs/hypersdk/state"
+
+	internalfees "github.com/ava-labs/hypersdk/internal/fees"
+)
+
+type Chain struct {
+	tracer                  trace.Tracer
+	config                  Config
+	metrics                 *chainMetrics
+	mempool                 Mempool
+	log                     logging.Logger
+	parser                  Parser
+	authVerificationWorkers workers.Workers
+	ruleFactory             RuleFactory
+	metadataManager         MetadataManager
+	balanceHandler          BalanceHandler
+	authVM                  AuthVM
+
+	lock       sync.Mutex
+	chainIndex ChainIndex
+	seen       *emap.EMap[*Transaction]
+}
+
+type ChainIndex interface {
+	GetExecutionBlock(ctx context.Context, blkID ids.ID) (*ExecutionBlock, error)
+	LastAcceptedBlock() *ExecutionBlock
+}
+
+type ChainBackend interface {
+	Tracer() trace.Tracer
+	Metrics() metrics.MultiGatherer
+	Mempool() Mempool
+	Logger() logging.Logger
+	Parser
+	RuleFactory() RuleFactory
+	MetadataManager() MetadataManager
+	BalanceHandler() BalanceHandler
+	AuthVerifiers() workers.Workers
+	GetAuthBatchVerifier(authTypeID uint8, cores int, count int) (AuthBatchVerifier, bool)
+	ChainIndex
+}
```

Comment:
It seems like this interface is being used just to hold a bunch of dependencies... if so I would prefer if we just individually passed them as args into `NewChain`.

Reply:
This is following the existing pattern where the VM is effectively the "god object" that holds all of these dependencies.

I'd prefer to keep it as is in this PR and break it up in a future PR, but can break it up to: `chain.New(vm.Tracer(), vm.Metrics(), vm.Mempool(), ..., vm)` in this PR if you prefer